XiEditor/CoreConnection.swift

```diff
@@ -84,14 +84,34 @@ class CoreConnection {
     func handleRaw(_ data: Data) {
         do {
             let json = try JSONSerialization.jsonObject(with: data, options: .allowFragments)
-            //print("got \(json)")
-            if !handleRpcResponse(json as AnyObject) {
+//            print("got \(json)")
+            handleRpc(json as Any)
+        } catch {
+            print("json error \(error.localizedDescription)")
+        }
+    }
+
+    /// handle a JSON RPC call. Determines whether it is a request, response or notifcation
+    /// and executes/responds accordingly
+    func handleRpc(_ json: Any) {
+        if let obj = json as? [String: Any], let index = obj["id"] as? Int {
+            if let result = obj["result"] { // is response
+                var callback: ((Any?) -> ())? = nil
+                queue.sync {
+                    callback = self.pending.removeValue(forKey: index)
+                }
+                callback?(result)
+            } else { // is request
                 DispatchQueue.main.async {
-                    self.callback(json as AnyObject)
+                    let result = self.callback(json as AnyObject)
+                    let resp = ["id": index, "result": result] as [String : Any?]
```

Comment:
Since we are not 100% JSON RPC spec complaint anyway, I have omitted the `"error"` property, because it is not used anywhere in the frontend-core communication, isn't it? However, I am of course fine with adding it.